src/hotspot/share/cds/archiveHeapWriter.cpp

```diff
@@ -0,0 +1,658 @@
+  ensure_buffer_space(1); // so that buffer_bottom() works
+}
+
+void ArchiveHeapWriter::ensure_buffer_space(size_t min_bytes) {
+  // We usually have very small heaps. If we get a huge one it's probably caused by a bug.
+  guarantee(min_bytes <= max_jint, "we dont support archiving more than 2G of objects");
+  _buffer->at_grow(to_array_index(min_bytes));
+}
+
+void ArchiveHeapWriter::copy_roots_to_buffer(GrowableArrayCHeap<oop, mtClassShared>* roots) {
+  Klass* k = Universe::objectArrayKlassObj(); // already relocated to point to archived klass
+  int length = roots != nullptr ? roots->length() : 0;
+  _heap_roots_word_size = objArrayOopDesc::object_size(length);
+  size_t byte_size = _heap_roots_word_size * HeapWordSize;
+  if (byte_size >= MIN_GC_REGION_ALIGNMENT) {
+    log_error(cds, heap)("roots array is too large. Please reduce the number of classes");
+    vm_exit(1);
+  }
+
+  maybe_fill_gc_region_gap(byte_size);
+
+  size_t new_top = _buffer_top + byte_size;
+  ensure_buffer_space(new_top);
+
+  HeapWord* mem = offset_to_buffered_address<HeapWord*>(_buffer_top);
+  memset(mem, 0, byte_size);
+  {
+    // This is copied from MemAllocator::finish
+    oopDesc::set_mark(mem, markWord::prototype());
+    oopDesc::release_set_klass(mem, k);
+  }
+  {
+    // This is copied from ObjArrayAllocator::initialize
+    arrayOopDesc::set_length(mem, length);
+  }
+
+  objArrayOop arrayOop = objArrayOop(cast_to_oop(mem));
+  for (int i = 0; i < length; i++) {
+    // Do not use arrayOop->obj_at_put(i, o) as arrayOop is outside of the real heap!
+    oop o = roots->at(i);
+    if (UseCompressedOops) {
+      * arrayOop->obj_at_addr<narrowOop>(i) = CompressedOops::encode(o);
+    } else {
+      * arrayOop->obj_at_addr<oop>(i) = o;
+    }
+  }
+  log_info(cds)("archived obj roots[%d] = " SIZE_FORMAT " bytes, klass = %p, obj = %p", length, byte_size, k, mem);
+
+  _heap_roots_bottom = _buffer_top;
+  _buffer_top = new_top;
+}
+
+void ArchiveHeapWriter::copy_source_objs_to_buffer(GrowableArrayCHeap<oop, mtClassShared>* roots) {
+  copy_source_objs_to_buffer_by_region(/*copy_open_region=*/true);
+  copy_roots_to_buffer(roots);
+  _open_top = _buffer_top;
+
+  // Align the closed region to the next G1 region
+  _buffer_top = _closed_bottom = align_up(_buffer_top, HeapRegion::GrainBytes);
+  copy_source_objs_to_buffer_by_region(/*copy_open_region=*/false);
+  _closed_top = _buffer_top;
+
+  log_info(cds, heap)("Size of open region   = " SIZE_FORMAT " bytes", _open_top   - _open_bottom);
+  log_info(cds, heap)("Size of closed region = " SIZE_FORMAT " bytes", _closed_top - _closed_bottom);
+}
+
+void ArchiveHeapWriter::copy_source_objs_to_buffer_by_region(bool copy_open_region) {
+  for (int i = 0; i < _source_objs->length(); i++) {
+    oop src_obj = _source_objs->at(i);
+    HeapShared::CachedOopInfo* info = HeapShared::archived_object_cache()->get(src_obj);
+    assert(info != nullptr, "must be");
+    if (info->in_open_region() == copy_open_region) {
```

Comment:
This trick `(false == false)` is `true` used to copy closed regions fooled me for a while. This was unusual for me, so pointing out here for anyone else reading it up. This would anyway go away in the future work that combines open and closed regions into a single one.